Stabilize `-Zcheck-cfg` as always enabled (#13571)
rust-lang/cargo: src/doc/src/reference/build-scripts.md

  compiler.
* [`cargo::rustc-cfg=KEY[="VALUE"]`](#rustc-cfg) --- Enables compile-time `cfg`
  settings.

> * [`cargo::rustc-check-cfg=CHECK_CFG`](#rustc-check-cfg) -- Enables compile-time

Reviewer: This doesn't enable compile-time checking but registers `cfg`s as valid for use with the compile-time checking

Author: Changed to `Register custom cfgs as expected for compile-time checking of configs`

The `rustc-cfg` instruction tells Cargo to pass the given value to the
[`--cfg` flag][option-cfg] to the compiler. This may be used for compile-time
detection of features to enable [conditional compilation]. Custom cfgs
must either be expected using the [`cargo::rustc-check-cfg`](#rustc-check-cfg)

> instruction or usage will need to allow the `unexpected_cfgs` lint to avoid

Reviewer 2: It will be great to have a link to [the lint rule](https://doc.rust-lang.org/nightly/rustc/lints/listing/warn-by-default.html#unexpected-cfgs).

Author: Sure, done.

The `rustc-cfg` instruction tells Cargo to pass the given value to the
[`--cfg` flag][option-cfg] to the compiler. This may be used for compile-time
detection of features to enable [conditional compilation]. Each custom cfgs
must **always** be declared using the [`cargo::rustc-check-cfg`](#rustc-check-cfg)

> instruction.

Author: Changed to:

> Custom cfgs must either be expected using the [`cargo::rustc-check-cfg`](#rustc-check-cfg) instruction or usage will need to allow the `unexpected_cfgs` lint to avoid unexpected cfgs warnings.

[conditional compilation]: ../../reference/conditional-compilation.md
[option-cfg]: ../../rustc/command-line-arguments.md#option-cfg

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

The `rustc-check-cfg` instruction tells Cargo to pass the given expected
configs to the [`--check-cfg` flag][option-check-cfg] to the compiler. This is
used for compile-time detection of unexpected conditional compilation configs.

You can use the instruction like this:

```rust,no_run
// build.rs
println!("cargo::rustc-check-cfg=cfg(foo)");
```

Note that it is expected to **always** pass the full list of custom configs
used in the package. The recommended way to achieve this is put all the

> check-cfg instructions at the top of your `build.rs`.

Reviewer: What do you mean by "full"?

Author: The complete list of all the custom cfgs and their values, even if some combinations are not technically possible.

Reviewer: maybe this could be reworded to "all possible `cfg`, regardless of what is currently active".

Also a bit unsure of the idea of "put all at the top".  I expect there would be cases where that'd be a terrible thing to do from a maintenance perspective.

Reviewer: > The "put all at the top" is mainly to suggest users to put all of the check-cfg instructions at the start of the build.rs main, as a way to make sure they are before anything else (in particular any conditions or whatever).

The `check-cfg` logic has a logical dependency on the `cfg` logic for generating the correct set of cfgs.  The further away they are, the more likely they are to get out of sync.  We should not be encouraging them to be further away, making them more likely to get out of sync.

(this is assuming multiple `check-cfg` directives can be given)

Author: I'm basing this recommendation on the fact that [`bootstrap`](https://github.com/rust-lang/rust/blob/d255c6a57c393db6221b1ff700daea478436f1cd/src/bootstrap/src/lib.rs#L74-L89) and [`libc`](https://github.com/rust-lang/libc/blob/a0f5b4b21391252fe38b2df9310dc65e37b07d9f/build.rs#L5-L30) both defines their custom check-cfgs in a list at the top of `build.rs` (or equivalent). It IMO makes it clear from the start what are the expected cfgs, even if the actual Cargo instructions is berried away.

> (this is assuming multiple check-cfg directives can be given)

yes they can

Author: I agree that it's not perfect, but given that we don't have a "support crate", it's the simplest recommendation I can think of.

Maybe we could recommend somethings similar to the `libc` setup, but I fear it will be too involved for users to just apply it and they would roll out a worst version of both recommendation.

Reviewer: I agree that having a helper library could make this simpler to verify their own output and I agree it would be too much to tell people to be like libc.

I disagree with keeping this suggestion though especially because we don't expect people to be like libc.

Author: I'm fine removing it but not without an alternative.

Since your worried about the instruction being too far apart, what about something like this?

> One way to achieve this, is to group the check-cfg and cfg instructions as much as
possible.
>
> <details>
> 
> ```rust,no_run
> // build.rs
> println!("cargo::rustc-check-cfg=cfg(foo, values(\"bar\"))");
> if foo_bar_condition {
>     println!("cargo::rustc-cfg=foo=\"bar\"");
> }
> ```
> 
> </details>

The example makes it clear about what needs to be done, can be generalized if desired and respect your desired to put them as close as possible.

Reviewer: This works.  I'd  also be fine with just not giving direction on this.

[conditional compilation]: ../../reference/conditional-compilation.md
[option-cfg]: ../../rustc/command-line-arguments.md#option-cfg

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

The `rustc-check-cfg` instruction tells Cargo to pass the given expected
configs to the [`--check-cfg` flag][option-check-cfg] to the compiler. This is
used for compile-time detection of unexpected [conditional compilation] configs.

You can use the instruction like this:

```rust,no_run
// build.rs
println!("cargo::rustc-check-cfg=cfg(foo, values(\"bar\"))");
```

> Be aware that the [`--check-cfg` flag][option-check-cfg] expected all possible
> configs, regardless of which is currently enabled to **always** be passed.

Reviewer: ```suggestion
Note that all possible
cfgs and values should be defined, regardless of what cfgs or values are currently active
```
There were some odd pieces of grammar (e.g. "expected" instead of "expects") but in general felt the structure was odd.

Author: It was indeed very odd. Fixed using your suggestion as baseline.

Note that I changed some terminology, "active" -> "enabled" and avoided the "cfg and values" as cfg generally means "key=value", and instead put the values warning in it's own sentence.

<details>

> Note that all possible cfgs should be defined, regardless of which cfgs are currently enabled. This includes all possible values of a given cfg name.

</details>

[conditional compilation]: ../../reference/conditional-compilation.md
[option-cfg]: ../../rustc/command-line-arguments.md#option-cfg

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

The `rustc-check-cfg` instruction tells Cargo to pass the given expected
configs to the [`--check-cfg` flag][option-check-cfg] to the compiler. This is

> used for compile-time detection of unexpected [conditional compilation] configs.

Reviewer: Unsure about the other ones but this feels backwards, starting with / focusing on the implementation (which people likely won't understand) and then saying why its useful.

Author: I can understand that it can feel backwards, personally it doesn't bother me.

Maybe it can be written like this; not sure if it's better.
```markdown
The `rustc-check-cfg` instruction is used for compile-time detection of unexpected
[conditional compilation] configs by telling Cargo to pass the given expected
configs to the [`--check-cfg` flag][option-check-cfg] to the compiler.
```

Reviewer: People scan documentation.  The start of the description will either have people skip over it completely or they can misunderstand the entire behavior based off of that scan.

This flipped the statement but the statement can easily be misread.  This is saying this is for compile-time detection, which can read like this enables compile time detection.  This is for defining keys/values for conditional compilation configs so they don't get reported as unexpected.

Author: I reworked this part of the documention to include add an explaination of what it's doing and why it is doing so.

It looks like this:

-----

### `cargo::rustc-check-cfg=CHECK_CFG`

This instruction is related to checking the conditional configurations of the
crate at compile time. That is, checking that every _reachable_ `#[cfg]` matches
the expected config names and values.

This `rustc-check-cfg` instruction increase the list of expected config names
and values by telling Cargo which custom configs should be expected in addition
to Cargo features and [well known cfgs][well-known-cfgs].

It is backed by the [`--check-cfg` flag][option-check-cfg] of the compiler.

-----

I don't know if we should put more emphasis into why it's useful (not that is very explicit either).

Let me know what you think.

Author: @epage Are you satisfied with the above changes?

Reviewer: Still feels like the lead-in gives the wrong impression about what this is about

How about

---

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

Add to the list of expected config names values that used when checking reachable `cfg` expressions.

For details on the syntax of `CHECK_CFG`, see rustc's [`--check-cfg` flag][option-check-cfg].  See also `<link to rustc lint>`.

The instruction can be used like this:

...

---

Reviewer 2: Ed's suggestion is exactly what I proposed in <https://github.com/rust-lang/cargo/pull/13571#discussion_r1585451270> and <https://github.com/rust-lang/cargo/pull/13571#discussion_r1585459006>. Let me resolve those two then :)

Author: Ok, I took your suggestion; just tweaked it for my non native english eyes.

------

Add to the list of expected configs the names and values used when checking the _reachable_
cfg expressions.

For details on the syntax of `CHECK_CFG`, see `rustc` [`--check-cfg` flag][option-check-cfg].
See also the [`unexpected_cfgs`][unexpected-cfgs] lint.

Reviewer: I think the change was

> Add to the list of expected **config names values that** used when checking reachable cfg expressions.

to

> Add to the list of expected **configs the names and values** used when checking the reachable
cfg expressions.

Re-arranging that to make clear the what the components are, you get "adding names and values to expected configs".  I find the concept of "expected configs" to be odd" and instead see this as us adding expected config names and values.

Of course mine had some grammatical mistakes.  Correcting for that, I propose:

> Add to the list of expected **config names and values that is** used when checking reachable cfg expressions.

Thoughts?

Author: While writing down my (overly long) response, I found myself switch back and forth between "this is fine" and I prefer "expected configs". But seeing that we already use "expected config names and values" is other places, I don't have an objection.

The only thing I still find a bit weird with your sentence is the *what?*, *what is being added to the list?*

Reviewer: For me, that is implied with the directive itself, `CHECK_CFG`.  If you want to be more explicit, that is fine.

[conditional compilation]: ../../reference/conditional-compilation.md
[option-cfg]: ../../rustc/command-line-arguments.md#option-cfg

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

This instruction is related to checking the conditional configurations of the
crate at compile time. That is, checking that every _reachable_ `#[cfg]` matches
the expected config names and values.

The `rustc-check-cfg` instruction adds to the list of expected config names
and values in addition to Cargo features and [well known cfgs][well-known-cfgs].

> It is backed by the [`--check-cfg` flag][option-check-cfg] of the compiler.

Reviewer 2: Instead, could we say that the syntax is what `--check-cfg` accepts, like other instructions do?

> The rustc-link-arg instruction tells Cargo to pass the [-C link-arg=FLAG option](https://doc.rust-lang.org/nightly/rustc/codegen-options/index.html#link-arg) to the compiler,

Author: See this [thread](https://github.com/rust-lang/cargo/pull/13571#discussion_r1569517868) with `@epage` about why it differ, but the basic idea was to put the Why? first and then the How?.

The instruction can be used like this:

```rust,no_run
// build.rs
println!("cargo::rustc-check-cfg=cfg(foo, values(\"bar\"))");
```

Note that all possible cfgs should be defined, regardless of which cfgs are
currently enabled. This includes all possible values of a given cfg name.

It is recommended to group the `cargo::rustc-check-cfg` and
[`cargo::rustc-cfg`][option-cfg] instructions as closely as possible in order to
avoid typos, missing check-cfg, stalled cfgs...

> <details>

Reviewer 2: This will be the first place we use `<details>` element. I don't object to the usage here. Just feel like this example is quite succinct and doesn't need to wrap in `<details>`.

The alternative idea is that we might want to expand this [Conditional compilation](http://127.0.0.1:3000/reference/build-script-examples.html#conditional-compilation) example instead.

Author: No opinion on it either, but I can't see the example (localhost url).

Reviewer 2: My bad 😅.
Here it is <https://doc.rust-lang.org/nightly/cargo/reference/build-script-examples.html#conditional-compilation>.

Author: Good point. I propose we do both, remove the `<details>` and update the build script example.

Author: Updated both.

<summary>

Example of using `cargo::rustc-check-cfg` and `cargo::rustc-cfg` together

</summary>

```rust,no_run
// build.rs
println!("cargo::rustc-check-cfg=cfg(foo, values(\"bar\"))");
if foo_bar_condition {
    println!("cargo::rustc-cfg=foo=\"bar\"");
}
```

</details>

> [well-known-cfgs]: ../../rustc/check-cfg.html#well-known-names-and-values

Reviewer 2: nit: for unnecessary consistency 😬

```suggestion
[well-known-cfgs]: ../../rustc/check-cfg.md#well-known-names-and-values
```

[conditional compilation]: ../../reference/conditional-compilation.md
[option-cfg]: ../../rustc/command-line-arguments.md#option-cfg

### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}

This instruction is related to checking the conditional configurations of the
crate at compile time. That is, checking that every _reachable_ `#[cfg]` matches
the expected config names and values.

> This `rustc-check-cfg` instruction increase the list of expected config names
> and values by telling Cargo which custom configs should be expected in addition
> to Cargo features and [well known cfgs][well-known-cfgs].

Reviewer 3: Just some editorial simplification.

```suggestion
The `rustc-check-cfg` instruction adds to the list of expected config names
and values in addition to Cargo features and [well known cfgs][well-known-cfgs].
```
